swift-synthesize-interface: Infer target triple

tshortli opened the pull request:
Infer the `-target` argument to `swift-synthesize-interface` to be the host triple when unspecified instead of emitting an error.

Resolves rdar://156353450.

tshortli committed:
swift-synthesize-interface: Infer target triple.

Infer the `-target` argument to `swift-synthesize-interface` to be the host
triple when unspecified instead of emitting an error.

Resolves rdar://156353450.

tshortli commented:
@swift-ci please smoke test

allevato reviewed (approved):
This makes sense to me as a nice QoL improvement.

FWIW, this logic was borrowed from swift-symbolgraph-extract (https://github.com/swiftlang/swift/blob/8803d8359c2cffbcde0c25cacd737929cfecaf94/lib/DriverTool/swift_symbolgraph_extract_main.cpp#L86); does it make sense to do the same thing there?

DougGregor reviewed (approved):
Thank you!

tshortli committed:
swift-symbolgraph-extract: Infer target triple.

Infer the `-target` argument to `swift-synthesize-interface` to be the host
triple when unspecified instead of emitting an error.

tshortli commented:
> This makes sense to me as a nice QoL improvement.
> 
> FWIW, this logic was borrowed from swift-symbolgraph-extract (
> 
> https://github.com/swiftlang/swift/blob/8803d8359c2cffbcde0c25cacd737929cfecaf94/lib/DriverTool/swift_symbolgraph_extract_main.cpp#L86
> 
> ); does it make sense to do the same thing there?

Sure! Done.

tshortli commented:
@swift-ci please smoke test

finagolfin commented on test/SymbolGraph/Relationships/MemberOf/Basic.swift:
```diff
@@ -1,7 +1,14 @@
+
+// Verify that -target can be inferred
+// RUN: %empty-directory(%t/Output)
+// RUN: %swift-symbolgraph-extract -module-name Basic -I %t -pretty-print -output-dir %t/Output
```
@tshortli, it appears [this line broke cross-compilation scenarios, like on Android](https://github.com/swift-android-sdk/swift-docker/actions/runs/16587224467/job/46914855078?pr=10):
```
2025-07-29T09:38:44.5213897Z RUN: at line 9: /home/build-user/build/swift-project/Ninja-Release/swift-linux-x86_64/bin/swift-symbolgraph-extract -module-name Basic -I /home/build-user/build/swift-project/Ninja-Release/swift-linux-x86_64/test-android-aarch64/SymbolGraph/Relationships/MemberOf/Output/Basic.swift.tmp -pretty-print -output-dir /home/build-user/build/swift-project/Ninja-Release/swift-linux-x86_64/test-android-aarch64/SymbolGraph/Relationships/MemberOf/Output/Basic.swift.tmp/Output
2025-07-29T09:38:44.5218326Z + /home/build-user/build/swift-project/Ninja-Release/swift-linux-x86_64/bin/swift-symbolgraph-extract -module-name Basic -I /home/build-user/build/swift-project/Ninja-Release/swift-linux-x86_64/test-android-aarch64/SymbolGraph/Relationships/MemberOf/Output/Basic.swift.tmp -pretty-print -output-dir /home/build-user/build/swift-project/Ninja-Release/swift-linux-x86_64/test-android-aarch64/SymbolGraph/Relationships/MemberOf/Output/Basic.swift.tmp/Output
2025-07-29T09:38:44.5221723Z <unknown>:0: error: module 'Basic' was created for incompatible target aarch64-unknown-linux-android: /home/build-user/build/swift-project/Ninja-Release/swift-linux-x86_64/test-android-aarch64/SymbolGraph/Relationships/MemberOf/Output/Basic.swift.tmp/Basic.swiftmodule
2025-07-29T09:38:44.5223384Z Error: Failed to load the module 'Basic'. Are you missing build dependencies or include/framework directories?
2025-07-29T09:38:44.5224081Z See the previous error messages for details. Aborting.
```
Are you seeing this error on other cross-compilation CI too?
